pkg/runtimecompiler/config/config.go

```diff
@@ -0,0 +1,85 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+// +build linux_bpf
+
+package config
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/cmd/system-probe/config"
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/ebpf"
+	netconfig "github.com/DataDog/datadog-agent/pkg/network/config"
+	secconfig "github.com/DataDog/datadog-agent/pkg/security/config"
+)
+
+const (
+	rcNS = "runtime_compiler_config"
+	spNS = "system_probe_config"
+	nwNS = "network_config"
+	rsNS = "runtime_security_config"
+)
+
+// Config stores all flags related to runtime compilation
+type Config struct {
```

Comment:
Should the configs https://github.com/DataDog/datadog-agent/pull/11760/files#diff-7a70d1343f42219c3b2c07d3bf90635f3d79b9e4fa7ed18ede433218e25e929eR75-R81 move here explicitly?

Reply:
So we have `ebpf.Config` nested multiple times here? i.e. it is part of `runtimecompiler.Config`, and e.g., `network.Config` which itself is nested in `runtimecompiler.Config`? 

I think it would be much clearer if we had another struct for the runtime compiler settings that are currently in `ebpf.Config` and that gets included in `runtimecompiler.Config`. I would also recommend you not include `network.Config` and `security.Config` in `runtimecompiler.Config`, and see if you could just have fields that are needed (not the whole structs). I would prefer you not use `ebpf.Config` at all here for these settings.

Reply:
What you're suggesting doesn't solve the main issue though - which is that the network tracer, security tracer, OOM kill probe & TCP queue length probe all need access to the runtime compiler settings for calls like this: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/network/tracer/connection/kprobe/tracer.go#L89

And since the OOM Kill probe and TCP queue length probes receive `ebpf.Config` as their main config, no matter how you slice it the runtime compiler settings will need to visible in some way from within `ebpf.Config`: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/collector/corechecks/ebpf/probe/oom_kill.go#L39-L40

We could have the runtime compiler settings exist within their own config which gets nested in `ebpf.Config`, but then from `network.Config` and `security.Config` we would have to access those settings through 2 layers of nesting. I don't really think that would be any clearer, especially since the runtime compilation settings currently exist within `ebpf.Config`, so it's a natural place for them to be.

On a different note - I can definitely remove `network.Config` and `security.Config` from `runtimecompiler.Config`, which also means that `ebpf.Config` wouldn't be nested multiple times. The only reason I had opted to have both those configurations in their entirety be fields in `runtimecompiler.Config` was so that when people add new runtime compilation assets in the future, they won't need to add settings for those assets in 2 separate places. But that's also not something I feel strongly about, so if you think it would be clearer to remove those fields I'm fine with that

Reply:
It sounds like there are two sets of configs -- one that is global and another that is module specific (for network/security, etc.). There is a singleton instance of the runtime module, so one option could be to expose the runtime compiler config as an exported property of that singleton for other modules to consume. 

I think it would be clearer to not nest the network and security configs -- they can be pretty big as well as muddling the context a little as well.

Reply:
That would be a perfect solution.... except for this annoying piece of code: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/runtimecompiler/runtimecompiler.go#L176-L178

Basically, the security probe's runtime-compiled constant fetcher dynamically generates c code which needs to be compiled, so we need to generate that code ourselves in order to compile it before the security probe runs (and then the security probe will later generate it again in order to hash the input & load the correct runtime-compiled output file). This creates a dependency between the compiler and the security probe/the security probe's `constantfetch` module, which means that we can't also have the inverse dependency where the security probe & constant fetcher need to read a config setting from the compiler.

The only way around this would be to have the functions related to generating the C code live inside their own module. This is possible, however these functions are very closely tied to the `constantfetch` and `probe` modules of the security probe, so this extraction makes those modules significantly less clear.

Reply:
It only seems to be using one config: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/security/probe/constantfetch/runtime_compiled.go#L109? I think you can make some modifications there to accommodate it reading from a global config object that has that config.

Reply:
Yep, `RuntimeCompilerOutputDir` is the only config setting which any of this applies to - I just didn't want to treat it differently from the other runtime compilation settings because I thought it was neater to keep all them all in the same place. I don't feel strongly about that though, so I'll just go ahead and move the other settings to `runtimecompiler.Config` to keep this discussion simpler

Reply:
Now, `RuntimeCompilerOutputDir` is the only runtime compilation related setting left in `ebpf.Config`. Since all the modules need to access this setting, our options are to leave it where it is, or to make a global variable (/global configuration structure) to hold it. Since runtime compilation is related to ebpf, I don't personally see an issue with runtime compilation settings living within the ebpf settings. For this reason, I don't think there's a strong enough case here to make it a global variable.

(FWIW, all of this discussion has no impact on customers, as either way these settings will all live under `runtime_compilation_config` in the `system-probe.yaml` file.)

Reply:
The new changes look great btw! 

I tend to see it as being left out there on its own so I would personally wonder why that is there and everything else in `runtimecompiler.Config`. What problems do you see with something like `runtimecompiler.RuntimeCompiler.Config` being used by the other modules to access all these settings?

Reply:
I don't personally see a way to get either the `constantfetch` or `probe` package to be able to read from `runtimecompiler.RuntimeCompiler.Config` neatly, because of the dependency cycle I mentioned above. Removing the dependency cycle would require restructuring the constant fetcher code in a way which I think would make it significantly less clear.

We could have all the other modules read from `runtimecompiler.RuntimeCompiler.Config` and do something different in the security module - specifically, we could pass `RuntimeCompilerOutputDir` to the security probe upon initialization and store the value within the probe itself. I don't personally like this solution very much, but regardless, my bigger concern with this pattern is that if we (or any other module) want to do the same thing as the security module is currently doing (aka generating code for compilation dyamically), we'll have import cycles to deal with ourselves. I'd prefer to stick to a model where the runtime compiler can access all the other modules if it needs to, and the other modules don't even need to know the compiler exists.

Reply:
@paulcacheux  That's definitely good to know, thanks!

@hmahmood  I changed the name of `RuntimeCompilerOutputDir` in an effort to make it more clear why it belongs in the eBPF config - hopefully that alleviates your concerns a little

Reply:
Sounds like with these sorts of requirements, we are going to have an issue with cyclic dependencies, especially with the global singleton we have for the runtime compiler. The cycle for the config you can likely solve by having a global in the `runtimecompiler/config` package. But it sounds like this will be an issue for other components of the runtime compiler module in the future. 

I'll leave it to you to decide if you want to move to using a global from the config package instead. For now, I think this'll do. Thanks for making all those changes!

Reply:
Sounds good. Thanks a bunch for the review Hasan!